File: leetcode/207.CourseSchedule.go
package leetcode

func canFinish(numCourses int, prerequisites [][]int) bool {
	mappings := make(map[int][]int)

	for i := 0; i < numCourses; i++ {
		mappings[i] = []int{}
	}

	for _, prerequisite := range prerequisites {
		course, pre := prerequisite[0], prerequisite[1]
		mappings[course] = append(mappings[course], pre)
	}

	set := make(map[int]struct{})
	var traverse func(course int) bool

	traverse = func(course int) bool {
		_, ok := set[course]
		if ok {
			return false
		}

		if len(mappings[course]) == 0 {
			return true
		}

		set[course] = struct{}{}
		for _, pre := range mappings[course] {
			if !traverse(pre) {
				return false
			}
		}

		delete(set, course)
		mappings[course] = []int{}

		return true
	}

	for i := 0; i < numCourses; i++ {
		if !traverse(i) {
			return false
		}
	}

	return true
}

File: leetcode/207.CourseSchedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  const map = [...new Array(numCourses)].reduce((acc, _, i) => {
    acc[i] = []
    return acc
  }, {} as Record<number, number[]>)

  for (const prerequisite of prerequisites) {
    const [course, pre] = prerequisite
    map[course].push(pre)
  }

  const set = new Set<number>()
  const traverse = (course: number) => {
    if (set.has(course)) return false
    if (map[course].length === 0) return true

    set.add(course)
    for (const pre of map[course]) {
      if (!traverse(pre)) return false
    }
    set.delete(course)
    map[course] = []
    return true
  }

  for (let i = 0; i < numCourses; i++) {
    if (!traverse(i)) return false
  }

  return true
}
console.log(canFinish(2, [[1, 0]]))
